milli/src/update/word_prefix_pair_proximity_docids.rs

```diff
@@ -169,89 +368,348 @@ impl<'t, 'u, 'i> WordPrefixPairProximityDocids<'t, 'u, 'i> {
+    writer: grenad::Writer<std::fs::File>,
 ) -> Result<()> {
-    *current_prefixes = match current_prefixes.take() {
-        Some(prefixes) if w2.starts_with(&prefixes[0]) => Some(prefixes),
-        _otherwise => {
-            write_prefixes_in_sorter(prefixes_cache, word_prefix_pair_proximity_docids_sorter)?;
-            prefix_fst_keys.iter().find(|prefixes| w2.starts_with(&prefixes[0]))
+    let file = writer.into_inner()?;
+    let reader = grenad::Reader::new(BufReader::new(file))?;
+    if database.is_empty(wtxn)? {
+        let mut out_iter = database.iter_mut::<_, ByteSlice, ByteSlice>(wtxn)?;
+        let mut cursor = reader.into_cursor()?;
+        while let Some((k, v)) = cursor.move_on_next()? {
+            // safety: the key comes from the grenad reader, not the database
+            unsafe { out_iter.append(k, v)? };
+        }
+    } else {
+        let mut cursor = reader.into_cursor()?;
+        while let Some((k, v)) = cursor.move_on_next()? {
+            database.put::<_, ByteSlice, ByteSlice>(wtxn, k, v)?;
         }
-    };
+    }
+    Ok(())
+}
 
-    if let Some(prefixes) = current_prefixes {
-        buffer.clear();
-        buffer.extend_from_slice(w1.as_bytes());
-        buffer.push(0);
-        for prefix in prefixes.iter() {
-            if prefix.len() <= max_prefix_length && w2.starts_with(prefix) {
-                buffer.truncate(w1.len() + 1);
-                buffer.extend_from_slice(prefix.as_bytes());
-                buffer.push(prox);
-
-                match prefixes_cache.get_mut(buffer.as_slice()) {
-                    Some(value) => value.push(data.to_owned()),
-                    None => {
-                        prefixes_cache.insert(buffer.clone(), vec![data.to_owned()]);
-                    }
+/** A prefix trie. Used to iterate quickly over the prefixes of a word that are
+within a set.
+
+## Structure
+The trie is made of nodes composed of:
+1. a byte character (e.g. 'a')
+2. whether the node is an end node or not
+3. a list of children nodes, sorted by their byte character
+
+For example, the trie that stores the strings `[ac, ae, ar, ch, cei, cel, ch, r, rel, ri]`
+is drawn below. Nodes with a double border are "end nodes".
+
+┌──────────────────────┐ ┌──────────────────────┐ ╔══════════════════════╗
+│          a           │ │          c           │ ║          r           ║
+└──────────────────────┘ └──────────────────────┘ ╚══════════════════════╝
+╔══════╗╔══════╗╔══════╗ ┌─────────┐  ╔═════════╗ ┌─────────┐ ╔══════════╗
+║  c   ║║  e   ║║  r   ║ │    e    │  ║    h    ║ │    e    │ ║    i     ║
+╚══════╝╚══════╝╚══════╝ └─────────┘  ╚═════════╝ └─────────┘ ╚══════════╝
+                         ╔═══╗ ╔═══╗                 ╔═══╗
+                         ║ i ║ ║ l ║                 ║ l ║
+                         ╚═══╝ ╚═══╝                 ╚═══╝
+*/
+#[derive(Default, Debug)]
+struct PrefixTrieNode {
+    children: Vec<(PrefixTrieNode, u8)>,
+    is_end_node: bool,
+}
```

Comment:
How much memory will this tree consume in normal use cases? I mean, will it be flush fast enough to not trigger a memory issue.

Reply:
It takes a completely negligible amount of memory and never gets flushed. Reasons:
1. it contains only the prefixes corresponding to more than 50 different words, which are rare
2. the maximum length of the prefixes is very short (typically 2 bytes only)